src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -41,54 +42,78 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_copy_ep_t, uct_ep_t);
                    (_iov)->length, (_remote_addr))
 
-#define UCT_CUDA_COPY_CHECK_AND_CREATE_STREAM(_iface, _id) \
-    if ((_iface)->stream[_id] == 0) { \
-        ucs_status_t __status; \
-        __status = UCT_CUDA_FUNC_LOG_ERR(cudaStreamCreateWithFlags(&(_iface)->stream[_id], \
-                                                                   cudaStreamNonBlocking)); \
-        if (UCS_OK != __status) { \
-            return UCS_ERR_IO_ERROR; \
-        } \
+static UCS_F_ALWAYS_INLINE cudaStream_t *
+uct_cuda_copy_get_stream(uct_cuda_copy_iface_t *iface,
+                         ucs_memory_type_t src_type,
+                         ucs_memory_type_t dst_type)
+{
+    cudaStream_t *stream;
+    ucs_status_t status;
+
+    stream = &iface->stream[src_type][dst_type];
+    if (*stream == 0) {
+        status = UCT_CUDA_FUNC_LOG_ERR(cudaStreamCreateWithFlags(stream,
+                                                                 cudaStreamNonBlocking));
+        if (status != UCS_OK ) {
+            stream = NULL;
+        }
     }
 
+    return stream;
+}
+
 static UCS_F_ALWAYS_INLINE ucs_status_t
-uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src, size_t length,
-                                   enum cudaMemcpyKind direction,
-                                   uct_cuda_copy_stream_t id,
-                                   uct_completion_t *comp)
+uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
+                                   size_t length, uct_completion_t *comp)
 {
     uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_cuda_copy_iface_t);
     uct_cuda_copy_event_desc_t *cuda_event;
     ucs_status_t status;
+    ucs_memory_info_t src_info;
+    ucs_memory_info_t dst_info;
+    cudaStream_t *stream;
+    ucs_queue_head_t *event_q;
 
     if (!length) {
         return UCS_OK;
     }
 
+    if (length < iface->config.detect_thresh) {
+        /* pick random mem_types */
+        /* TODO: how expensive is random number generation wrt pointer query */
+        ucs_rand_range(UCS_MEMORY_TYPE_HOST, (UCS_MEMORY_TYPE_LAST - 1),
+                       (int*)&src_info.type);
+        ucs_rand_range(UCS_MEMORY_TYPE_HOST, (UCS_MEMORY_TYPE_LAST - 1),
+                       (int*)&dst_info.type);
+
+    } else {
+        uct_cuda_copy_memory_detect(iface, src, length, &src_info);
+        uct_cuda_copy_memory_detect(iface, dst, length, &dst_info);
+    }
+
+    stream  = uct_cuda_copy_get_stream(iface, src_info.type, dst_info.type);
+    event_q = &iface->outstanding_event_q[src_info.type][dst_info.type];
+
     cuda_event = ucs_mpool_get(&iface->cuda_event_desc);
     if (ucs_unlikely(cuda_event == NULL)) {
         ucs_error("Failed to allocate cuda event object");
         return UCS_ERR_NO_MEMORY;
     }
 
-    UCT_CUDA_COPY_CHECK_AND_CREATE_STREAM(iface, id);
+    status = UCT_CUDA_FUNC_LOG_ERR(cudaMemcpyAsync(dst, src, length, cudaMemcpyDefault,
```

Comment:
does using cudaMemcpyDefault here add any overhead?

Reply:
Using `cudaMemcpyDefault` and say `cudaMemcpy*To*` didn't show a bandwidth difference. I've not checked if there is difference in latency but for the current use case, it should be ok if there was a small difference.